By default, use MKL as virtual provider for blas/lapack/fftw with Intel compilers (classic and llvm-based/oneapi); update site configs to revert to openblas/fftw as needed; skip wgrib2 with Intel oneapi; bump odc to 1.5.2

climbfuji opened the pull request:
### Summary

Describe the changes made in this PR and why they are needed.

1. Unrelated change but needed for `gcc@13` support: bump `odc` from `1.4.6` to `1.5.2`.

2. Split configs/common/packages.yaml into a compiler-independent `configs/common/packages.yaml` and compiler-dependent `configs/common/packages_${COMPILER}.yaml`; use `openblas` and `fftw` as virtual providers for `blas`, `lapack`, `fftw-api` with `gnu@` and `apple-clang@`; use `intel-oneapi-mkl` with `intel@` and `oneapi@`.

3. Site config updates for all sites: split `packages.yaml` into `packages_${COMPILER}.yaml` and add Intel MKL as external package for `intel@` and `oneapi@` compilers. Please follow the examples for blackpearl, narwhal, nautilus. Note that updating the site config does **not** imply testing the update (see section "Testing" below for which tests where done).

 **Update 2024/08/09:** Certain site configs were modified to by default retain the `openblas/fftw` configuration with Intel - see list below. Steps to switch to the new default MKL configuration are documented in each site's `packages_*.yaml`.

List of sites:
  - [x] blackpearl (@climbfuji) - uses MKL w/ Intel
  - [x] acorn (@climbfuji) - keeps openblas w/ Intel
  - [x] atlantis (@climbfuji) - uses MKL w/ Intel
  - [x] aws-pcluster (@climbfuji) - keeps openblas w/ Intel
  - [x] casper (@climbfuji) - no changes needed, only has GNU
  - [x] derecho (@climbfuji) - keeps openblas w/ Intel
  - [x] discover-scu16 (@climbfuji) - uses MKL w/ Intel
  - [x] discover-scu17 (@climbfuji) - uses MKL w/ Intel
  - [x] gaea-c5 (@climbfuji) - keeps openblas w/ Intel
  - [x] gaea-c6 (@climbfuji) - keeps openblas w/ Intel
  - [x] hera (@climbfuji) - keeps openblas w/ Intel
  - [x] hercules (@climbfuji) - keeps openblas w/ Intel
  - [x] jet (@climbfuji) - keeps openblas w/ Intel
  - [x] narwhal (@climbfuji) - uses MKL w/ Intel; includes an update of the Intel compiler in the site config
  - [x] nautilus (@climbfuji) - uses MKL w/ Intel
  - [x] NOAA ParallelWorks (@climbfuji) - keeps openblas w/ Intel
  - [x] orion (@climbfuji) - keeps openblas w/ Intel
  - [x] s4 (@climbfuji) - keeps openblas w/ Intel
  - [x] sandy (@climbfuji) - no changes needed, only has GNU

5. Corresponding documentation updates.

### Testing

@fmahebert @srherbener @RatkoVasic-NOAA @natalie-perlin @AlexanderRichert-NOAA I "assigned" this PR to you in case you want to test the updated site configs (see list above) on the system(s) that you are responsible for - this is optional, because we'll go through all platforms in a few weeks anyway when we roll out spack-stack-1.8.0.

- [x] Built unified environment on blackpearl with gcc@13.3.0 and oneapi@2024.1.2 (@climbfuji)
- [x] Built neptune standalone environment on Narwhal with gcc@10.3.0 and intel@2021.10.0 (@climbfuji)
- [x] Built unified environment on Orion with intel@2021.9.0, with openblas (default per site config) and with MKL (manually changed) (@climbfuji)
- [x] CI

### Applications affected

All

### Systems affected

All

### Dependencies

n/a

### Issue(s) addressed

Resolves https://github.com/JCSDA/spack-stack/issues/759

### Checklist
- [x] This PR addresses one issue/problem/enhancement, or has a very good reason for not doing so.
- [x] These changes have been tested on the affected systems and applications.
- [ ] ~~All dependency PRs/issues have been resolved and this PR can be merged.~~

climbfuji committed:
Create packages_{apple-clang,gcc,intel,oneapi}.yaml and move compiler-specific settings out of packages.yaml

climbfuji committed:
Update blackpearl site config: remove oneapi settings (now in common packages), add mirror

climbfuji committed:
Update .gitmodules and submodule pointer for spack for code review and testing

climbfuji committed:
Temporarily remove wgrib2 requirement from various virtual packages for '%oneapi'

climbfuji committed:
First pass to update configs/sites/tier1/narwhal

climbfuji committed:
Update formatting in configs/common/packages_*.yaml

climbfuji commented on configs/common/packages.yaml:
```diff
@@ -1,267 +1,266 @@
 # Pin versions and specs when building packages
```
Please hide whitespace changes to see the actual diffs in this file (I adjusted the intent)

climbfuji committed:
Update nautilus site config: use packages_COMPILER.yaml

climbfuji committed:
Remove unused external intel-oneapi-tbb from Narwhal site config

climbfuji committed:
Update CI workflows to use Intel MKL with Intel and OneAPI compilers

climbfuji committed:
Merge branch 'develop' of https://github.com/jcsda/spack-stack into feature/oneapi_intel_use_mkl

climbfuji committed:
Update submodule pointer for spack

climbfuji committed:
Add libfabric to LD_LIBRARY_PATH in narwhal compiler config

climbfuji committed:
Bump odc to 1.5.2 in configs/common/packages.yaml and configs/containers/specs/jedi-ci.yaml

climbfuji committed:
Merge branch 'develop' of https://github.com/jcsda/spack-stack into feature/oneapi_intel_use_mkl

climbfuji committed:
For Intel Classic, pin gettext to 0.21.1 since 0.22.5 doesn't build

climbfuji committed:
Revert .gitmodules and update submodule pointer for spack

climbfuji committed:
Update atlantis site config

climbfuji committed:
Update S4 site config

climbfuji committed:
Update Orion site config

climbfuji committed:
Work in progress: update doc/source/NewSiteConfigs.rst and doc/source/PreConfiguredSites.rst

climbfuji committed:
Update Hercules site config

AlexanderRichert-NOAA commented:
I thought the plan was that for 1.8.0 we have a second, MKL-based Intel stack on some systems so developers can see if they like it before we go all in with it. Is that still the plan? I'm still concerned this could become a source of numerical differences that will need to be reconciled down the road, and I'm not sure I see the downside of continuing to use OpenBLAS, FFTW, etc., unless there's some really big performance difference.

climbfuji commented:
> I thought the plan was that for 1.8.0 we have a second, MKL-based Intel stack on some systems so developers can see if they like it before we go all in with it. Is that still the plan? I'm still concerned this could become a source of numerical differences that will need to be reconciled down the road, and I'm not sure I see the downside of continuing to use OpenBLAS, FFTW, etc., unless there's some really big performance difference.

I get the point that there may be b4b differences. But all we are doing here is going back to how things were in hpc-stack/jedi-stack. If you really want Intel classic to use openblas and fftw, then we can make that work on a per-site basis. It's then up to the maintainers on these systems to manually change the settings in order to create an MKL-based stack with Intel, in addition to the default openblas stack.

climbfuji committed:
[skip ci] TEST forcing orion intel config to use openblas and fftw instead of intel-oneapi-mkl

climbfuji committed:
Update submodule pointer for spack

climbfuji committed:
Merge branch 'develop' of https://github.com/jcsda/spack-stack into feature/oneapi_intel_use_mkl

climbfuji committed:
[skip ci]
TEST forcing orion intel config to use openblas and fftw instead of intel-oneapi-mkl part 2

climbfuji commented:
@AlexanderRichert-NOAA Please have a look at the updated Orion site config. I tested this just now and the intel build also concretizes with openblas and fftw. If this looks ok to you, I will apply the same settings for Hercules and S4. The other platforms I covered are either NRL systems that always used mkl, or they use gnu only (casper).

For all remaining platforms one just has to follow the orion example.

climbfuji committed:
Update S4 and Hercules site configs to use openblas and fftw with Intel

AlexanderRichert-NOAA commented:
Thanks. I'll check it out on a couple systems. Do you have a full build on, say, Hera that I could test UFS Weather Model RTs against?

climbfuji commented:
> Thanks. I'll check it out on a couple systems. Do you have a full build on, say, Hera that I could test UFS Weather Model RTs against?

I haven't touched the hera site config at all (only the ones ticked in the list at the top of the PR page), building on Orion now.

AlexanderRichert-NOAA commented:
Got it. Do you have the MKL-based stack on Hercules, and if so where?

climbfuji commented:
> Got it. Do you have the MKL-based stack on Hercules, and if so where?

None yet. I'll do this on Orion (because this is where I started). Currently building the openblas-fftw-stack, and after that the mkl stack. Should be ready on Monday.

climbfuji commented:
@AlexanderRichert-NOAA I have two environments on Orion:
```
/work2/noaa/jcsda/dheinzel/spack-stack-feature-oneapi_intel_use_mkl/envs/ue-intel-2021.9.0/install/modulefiles/Core
```
and
```
/work2/noaa/jcsda/dheinzel/spack-stack-feature-oneapi_intel_use_mkl/envs/ue-intel-2021.9.0-mkl/install/modulefiles/Core
```

climbfuji committed:
Merge branch 'develop' of https://github.com/jcsda/spack-stack into feature/oneapi_intel_use_mkl

climbfuji commented:
@AlexanderRichert-NOAA Here is a paper from Intel demonstrating better performance with MKL: https://www.intel.com/content/www/us/en/developer/articles/technical/performance-comparison-of-openblas-and-intel-math-kernel-library-in-r.html

climbfuji commented:
@AlexanderRichert-NOAA I tried using the current ufs-weather-model develop branch with this spack-stack PR and the two test installs on Orion noted above. I tried the `cpld_control_p8_mixedmode_intel` regression test. Both the openblas and the mkl builds segfault after/in the ww3 initialization:
```
> cat out
...
180:        Wave model ...
180:  WW3 log written to /work/noaa/stmp/dheinzel/stmp/dheinzel/FV3_RT/rt_2304865/cpl
180:  d_control_p8_mixedmode_intel_test_mkl/./log.ww3

> cat err
150: WARNING from PE     0: Unused line in INPUT/MOM_input : ODA_VINC_VAR = 'v_inc'
150:
150:
150: WARNING from PE     0: Unused line in INPUT/MOM_input : ODA_INCUPD_NHOURS = 6
150:
 77: Abort(805361423) on node 77 (rank 77 in comm 0): Fatal error in PMPI_Init_thread: Other MPI error, error stack:
 77: MPI_Init_thread(307): Cannot call MPI_INIT or MPI_INIT_THREAD more than once
```
I checked and esmf is linked static, mapl shared. I don't know why/how that changed. Remember that spack-stack develop uses newer ESMF and MAPL versions, and I also had to put this workaround in the ufs-weather-model top-level CMakeLIsts.txt to be able to run cmake and make:
```
diff --git a/CMakeLists.txt b/CMakeLists.txt
index e5fdd1e8..5c7a974a 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -148,6 +148,9 @@ endif()

 find_package(NetCDF 4.7.4 REQUIRED C Fortran)
 find_package(ESMF 8.3.0 MODULE REQUIRED)
+if (NOT TARGET ESMF::ESMF)
+       add_library(ESMF::ESMF ALIAS esmf)
+endif ()
 if(FMS)
   find_package(FMS 2022.04 REQUIRED COMPONENTS R4 R8)
   if(APP MATCHES "^(HAFSW)$")
```

Will try to force static mapl linking tomorrow.

climbfuji committed:
Update site configs for Hera and Jet (use openblas for now with Intel)

climbfuji committed:
Merge branch 'develop' of https://github.com/jcsda/spack-stack into feature/oneapi_intel_use_mkl

climbfuji commented:
@AlexanderRichert-NOAA Update. I rebuilt mapl as static libraries and linked against that version. With both MKL and blas, the ufs-weather-model still aborts with the same error and in the same place as described above. I am certain this is unrelated to the changes in this PR, it must have something to do with the update of some packages from spack-stack-1.6.0 to spack-stack-develop.

climbfuji committed:
Update site configs for aws-pcluster, derecho, discover-scu16, discover-scu17

climbfuji commented on .github/workflows/ubuntu-ci-x86_64-intel.yaml:
```diff
@@ -96,6 +96,17 @@ jobs:
+          echo "    externals:" >> ${SPACK_SYSTEM_CONFIG_PATH}/packages.yaml
+          echo "    - spec: intel-oneapi-mkl@2023.2.0%intel@2021.10.0" >> ${SPACK_SYSTEM_CONFIG_PATH}/packages.yaml
+          echo "      prefix: /opt/intel/oneapi" >> ${SPACK_SYSTEM_CONFIG_PATH}/packages.yaml
+          #echo "  intel-oneapi-runtime:" >> ${SPACK_SYSTEM_CONFIG_PATH}/packages.yaml
```
Remove intel-oneapi-runtime? Not needed for classic Intel, or?

climbfuji committed:
Update site configs for acorn, gaea-c5, gaea-c6

climbfuji committed:
Update site configs for noaa-{aws,azure,gcloud}

climbfuji committed:
Final update of doc/source/PreConfiguredSites.rst

climbfuji commented:
> @AlexanderRichert-NOAA Update. I rebuilt mapl as static libraries and linked against that version. With both MKL and blas, the ufs-weather-model still aborts with the same error and in the same place as described above. I am certain this is unrelated to the changes in this PR, it must have something to do with the update of some packages from spack-stack-1.6.0 to spack-stack-develop.

@AlexanderRichert-NOAA Yet another update. I compiled the spack-stack develop unified environment, then ufs-weather-model and ran the same test as above. It failed in the same place. Therefore, the problem is not related to this PR and it shouldn't be held up by that.

AlexanderRichert-NOAA commented:
Have you run other UWM RTs with it?

climbfuji commented:
> Have you run other UWM RTs with it?

I haven't.

climbfuji commented:
One thing I had to do in order to compile was:
```
index e5fdd1e8..5c7a974a 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -148,6 +148,9 @@ endif()

 find_package(NetCDF 4.7.4 REQUIRED C Fortran)
 find_package(ESMF 8.3.0 MODULE REQUIRED)
+if (NOT TARGET ESMF::ESMF)
+       add_library(ESMF::ESMF ALIAS esmf)
+endif ()
 if(FMS)
   find_package(FMS 2022.04 REQUIRED COMPONENTS R4 R8)
   if(APP MATCHES "^(HAFSW)$")
diff --git a/tests/logs/RegressionTests_orion.log b/tests/logs/RegressionTests_orion.log
index b08d50a9..eee3eda5 100644
```
This is because of the findESMF mismatches etc. discussed elsewhere. I don't think this is the cause of the problem, though ...

I'll run one basic regression test next (not sure when I get to it)

climbfuji commented:
@AlexanderRichert-NOAA control_c48 runs with spack-stack-dev and with this branch. The openblas config (default as per this PR) produces b4b identical results for control_c48 than spack-stack-dev. The mkl run is still stuck in the queue (orion had a power outage).

See https://github.com/ufs-community/ufs-weather-model/issues/2399 where Dusan reports that he gets the same errors I had above with the coupled run when using esmf 8.6.1 and mapl 2.46.2 in spack-stack-1.6.0.

I think at this point there is no reason to hold up this PR.

climbfuji commented:
Thanks @AlexanderRichert-NOAA

climbfuji commented:
@AlexanderRichert-NOAA I know this has been merged already, but for the sake of completeness: the control_c48 run with MKL is b4b identical to the openblas run. Of course, this may not be the case for the fully coupled model, but at least for atm standalone atmosphere with some stochastics (cellular automata) it is.
